### PR TITLE
add support for request_get_status_any/all/some

### PR DESCRIPTION
**Pull request by edgargabriel**
Add support for the new ```MPI_Request_get_status_any/all/some``` functions introduced with MPI 4.1

This PR contains the C implementation for the functions as well as the man-pages. 
There are however also some things missing:

- in the C code generation python scripts, we need to introduce the ability to create ```const MPI_Request[]```, (I have currently removed the ```const``` from the function definitions in mpi.h.in in order to make the code compile).
- Fortran interfaces

### CODE REVIEW

**Comment by hppritcha** [comment]
Let me know if you want help with the python binding infrastructure.

**Comment by edgargabriel** [comment]
> Let me know if you want help with the python binding infrastructure.

@hppritcha yes, I would appreciate some help on that front (and the same applies for the fortran bindings, even if its just a conversation that I understand what needs to be done).

File: ompi/mpi/c/request_get_status_all.c.in [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+/* Non blocking test for the request status. Upon completion, the request will
+ * not be freed (unlike the test function). A subsequent call to test, wait
+ * or free should be executed on the request.
+ */
+PROTOTYPE ERROR_CLASS request_get_status_all(INT count, REQUEST_INOUT requests:count, INT_OUT flag,
+                                             STATUS_OUT statuses:count)
+{
+#if OPAL_ENABLE_PROGRESS_THREADS == 0
+    int do_it_once = 0;
```
[the remark this answers is not in the document]
Comment (edgargabriel) [reply ompi/mpi/c/request_get_status_all.c.in]: I made the change. The main reason was because the code was based on request_get_status.c.in, which does it this way :-)

**Comment by hppritcha** [comment]
@edgargabriel added new data type to the bindings code to add this new use of ```const MPI_Request reqs[]``` in the standard.

**Comment by edgargabriel** [comment]
> note that mpi4py is probably not testing this since it keys off our MPI_Get_version return values for major/minor. @edgargabriel have you tested mpi4py with VERSION faked to return 4.1?

faking MPI 4.1. for mpi4py doesn't work unfortunately, we fail the compilation of mpi4py due to other missing function (MPI_Buffer_flush etc.). I will write some tests for the new functions.

**Comment by hppritcha** [comment]
@edgargabriel i pushed the remainder of the fortran related files to the PR - hopefully.

**Comment by edgargabriel** [comment]
> @edgargabriel i pushed the remainder of the fortran related files to the PR - hopefully.

@hppritcha  thank you! I hope to finish up adding tests to the ibm testsuite for the new function this weekend, and hopefully next week we could merge the PR if all tests pass.

**Comment by edgargabriel** [comment]
I added some tests for request_get_status_any/all/some to the ompi-tests private testsuite, and they pass. This is however only testing the C-interfaces of the function. The tests are also probably not entirely bullet-proof in terms of that they might miss some corner cases, but the base function is shown to be working.

File: ompi/mpi/bindings/bindings.py [hunk 2]
```diff
@@ -37,12 +37,16 @@ def main():
...
+    parser_code.add_argument('--lang', choices=('fortran', 'c', 'f90'),
                              help='language to generate (only for code subparser)')
+    parser_code.add_argument('--fort-std', choices=('f90', 'f08'),
+                             help='fortran standard to use for fortran module generation')
     # Handler for generating the Fortran interface files
     parser_interface = subparsers_fortran.add_parser('interface',
                                                      help='generate Fortran interface specifications')
     parser_interface.set_defaults(handler=lambda args, out: fortran.generate_interface(args, out))
+    parser_interface.add_argument('--fort-std', choices=('f90', 'f08'),
+                             help='fortran standard to use for fortran module generation')
```
Comment (jsquyres) [inline ompi/mpi/bindings/bindings.py]: @hppritcha I see `--fort-std` added in both `parser_code` and `parser_interface` -- is that intentional?

File: ompi/mpi/bindings/bindings.py [hunk 3]
```diff
@@ -37,12 +37,16 @@ def main():
     subparsers_fortran = parser_fortran.add_subparsers()
     parser_code = subparsers_fortran.add_parser('code', help='generate binding code')
     parser_code.set_defaults(handler=lambda args, out: fortran.generate_code(args, out))
-    parser_code.add_argument('--lang', choices=('fortran', 'c'),
+    parser_code.add_argument('--lang', choices=('fortran', 'c', 'f90'),
```
Comment (jsquyres) [inline ompi/mpi/bindings/bindings.py]: @hppritcha What is the `f90` choice for?  I don't see that being checked for in the python code -- am I missing it?

File: ompi/mpi/c/request_get_status_all.c.in [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+ recheck_request_status:
+#endif
+
+    opal_atomic_mb();
+    int i;
+    all_done = true;
+    for (i = 0; i < count; i++) {
+        one_done = false;
```
Comment (jsquyres) [inline ompi/mpi/c/request_get_status_all.c.in]: How is `one_done` ever set to true?  I think the "if" conditional below is incorrect.

File: ompi/mpi/c/request_get_status_all.c.in [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    if (OPAL_UNLIKELY(0 == count)) {
+        *flag = true;
+        return MPI_SUCCESS;
+    }
+
+    bool all_done;
+    bool one_done;
+
+#if OPAL_ENABLE_PROGRESS_THREADS == 0
+    int do_it_once = 0;
```
Comment (jsquyres) [inline ompi/mpi/c/request_get_status_all.c.in]: Is there a reason to make this an `int` instead of `bool`?

File: docs/man-openmpi/man3/MPI_Request_get_status_all.3.rst [hunk 6]
```diff
@@ -0,0 +1,48 @@
+.. _mpi_request_get_status_all:
+
+
+MPI_Request_get_status_all
+==========================
+
+.. include_body
+
+:ref:`MPI_Request_get_status_all` |mdash| Access information associated with a
+request without freeing the request.
```
Comment (jsquyres) [inline docs/man-openmpi/man3/MPI_Request_get_status_all.3.rst]: You're access information with all active requests in an array -- not just *a* request.

File: docs/man-openmpi/man3/MPI_Request_get_status_all.3.rst [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+DESCRIPTION
+-----------
+
+:ref:`MPI_Request_get_status_all` sets ``flag = true`` if all
+operations associated with *active* handles in the array have completed.
+In this case, each status entry that corresponds to an active request
+is set to the status of the corresponding operation. It
+does not deallocate or deactivate the request; a subsequent call to
+test, wait, or free should be executed with each of those requests.
```
Comment (jsquyres) [inline docs/man-openmpi/man3/MPI_Request_get_status_all.3.rst]: any of the MPI test, wait, or free routines ...

(note: all of my comments on this particular man page generally apply to all 3 of the man pages)

File: docs/man-openmpi/man3/MPI_Request_get_status_all.3.rst [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+Each status entry that corresponds to a null or inactive handle is set
+to empty.  Otherwise, ``flag = false`` is returned and the values of the
+status entries are undefined.
+
+If your application does not need to examine the *status* field, you can
+save resources by using the predefined constant ``MPI_STATUS_IGNORE`` as a
+special value for the ``array_of_statuses`` argument.
```
Comment (jsquyres) [inline docs/man-openmpi/man3/MPI_Request_get_status_all.3.rst]: Why would you call `MPI_Request_get_status_all` with `MPI_STATUSES_IGNORE`?  That doesn't make a lot of sense.

Is it even valid to call `MPI_Request_get_status_all` with `MPI_STATUSES_IGNORE`?

File: docs/man-openmpi/man3/MPI_Request_get_status_all.3.rst [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+is set to the status of the corresponding operation. It
+does not deallocate or deactivate the request; a subsequent call to
+test, wait, or free should be executed with each of those requests.
+
+Each status entry that corresponds to a null or inactive handle is set
+to empty.  Otherwise, ``flag = false`` is returned and the values of the
+status entries are undefined.
+
+If your application does not need to examine the *status* field, you can
+save resources by using the predefined constant ``MPI_STATUS_IGNORE`` as a
```
Comment (jsquyres) [inline docs/man-openmpi/man3/MPI_Request_get_status_all.3.rst]: Did you mean `MPI_STATUSES_IGNORE`?

File: ompi/mpi/c/request_get_status_all.c.in [hunk 10]
```diff
@@ -54,7 +54,7 @@ PROTOTYPE ERROR_CLASS request_get_status_all(INT count, REQUEST_INOUT requests:c
         if (NULL == requests) {
             rc = MPI_ERR_REQUEST;
         } else {
-            if(!ompi_request_check_same_instance(requests, count) ) {
+            if(!ompi_request_check_same_instance((MPI_Request *)requests, count) ) {
```
Comment (jsquyres) [inline ompi/mpi/c/request_get_status_all.c.in:61]: Why cast?  Is there a reason the 1st arg to `ompi_request_check_same_instance` isn't `const`?

File: ompi/mpi/fortran/use-mpi-f08/request_get_status_all.c.in [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+    for (i = 0; i < OMPI_FINT_2_INT(*count); ++i) {
+        c_req[i] = PMPI_Request_f2c(array_of_requests[i]);
+    }
+
+    c_ierr = @INNER_CALL@(OMPI_FINT_2_INT(*count), c_req, &c_flag, c_status);
+    if (NULL != ierr) *ierr = OMPI_INT_2_FINT(c_ierr);
+
+    if (MPI_SUCCESS == c_ierr) {
+        *flag = OMPI_INT_2_FINT(c_flag);
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/request_get_status_all.c.in:57]: `flag` is a LOGICAL.

File: ompi/mpi/fortran/use-mpi-f08/request_get_status_any.c.in [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    c_ierr = @INNER_CALL@(OMPI_FINT_2_INT(*count), c_req,
+                          &c_indx,
+                          &c_flag,
+                          &c_status);
+    if (NULL != ierr) *ierr = OMPI_INT_2_FINT(c_ierr);
+
+    if (MPI_SUCCESS == c_ierr) {
+
+        *indx = OMPI_INT_2_FINT(c_indx);
+        *flag = OMPI_INT_2_FINT(c_flag);
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/request_get_status_any.c.in:61]: `flag` is a LOGICAL.

File: ompi/mpi/fortran/mpif-h/request_get_status_some_f.c [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+    OMPI_ARRAY_FINT_2_INT_ALLOC(array_of_indices, *incount);
+    c_ierr = PMPI_Request_get_status_some(OMPI_FINT_2_INT(*incount), c_req,
+                                          OMPI_SINGLE_NAME_CONVERT(outcount),
+                                          OMPI_ARRAY_NAME_CONVERT(array_of_indices),
+                                          c_status);
+    if (NULL != ierr) *ierr = OMPI_INT_2_FINT(c_ierr);
+
+    if (MPI_SUCCESS == c_ierr) {
+        OMPI_SINGLE_INT_2_FINT(outcount);
+        OMPI_ARRAY_INT_2_FINT(array_of_indices, *incount);
```
Comment (jsquyres) [inline ompi/mpi/fortran/mpif-h/request_get_status_some_f.c]: Should this be *outcount?

File: ompi/mpi/fortran/use-mpi-f08/request_get_status_any.c.in (hunk 12 above)
Comment (hppritcha) [reply ompi/mpi/fortran/use-mpi-f08/request_get_status_any.c.in:61]: nope it isn't there.  See how the inner fortran binding works once you build the code out.

File: ompi/mpi/fortran/mpif-h/request_get_status_some_f.c (hunk 13 above)
Comment (hppritcha) [reply ompi/mpi/fortran/mpif-h/request_get_status_some_f.c]: oh yes.  it should be *outcount

File: ompi/mpi/fortran/use-mpi-f08/request_get_status_all.c.in (hunk 11 above)
Comment (hppritcha) [reply ompi/mpi/fortran/use-mpi-f08/request_get_status_all.c.in:57]: same below.  you need to look at the generated fortran code to see how this works.

File: ompi/mpi/bindings/bindings.py (hunk 3 above)
Comment (hppritcha) [reply ompi/mpi/bindings/bindings.py]: its implied in the check around line 297 in fortran.py but i'll make it explicit.  its suppose to mean generate the "f90" aka use mpi fortran interfaces.

File: ompi/mpi/bindings/bindings.py (hunk 2 above)
Comment (hppritcha) [reply ompi/mpi/bindings/bindings.py]: yes it is.  i can look if it makes sense to put in the higher level "fortran parser".

File: ompi/mpi/bindings/bindings.py (hunk 3 above)
Comment (hppritcha) [reply ompi/mpi/bindings/bindings.py]: actually i think that f90 here was an add on for future work.  it will be removed from this PR

File: ompi/mpi/c/request_get_status_all.c.in (hunk 10 above)
Comment (edgargabriel) [reply ompi/mpi/c/request_get_status_all.c.in:61]: These three functions (request_get_status_all/any/some) are the first ones that define the requests as `const`, all other interfaces (e.g. Waitall/Waitany etc.) do not have the `const` keyword. In addition, this function was not introduced in this pr, so if we decide to make a change it, I would suggest to do that in a separate PR.

File: ompi/mpi/c/request_get_status_all.c.in (hunk 5 above)
Comment (edgargabriel) [reply ompi/mpi/c/request_get_status_all.c.in]: changed to bool.

File: ompi/mpi/c/request_get_status_all.c.in (hunk 4 above)
Comment (edgargabriel) [reply ompi/mpi/c/request_get_status_all.c.in]: completely removed the one_done variable. It was a remainder of the first algorithm to implement the function, is not really required meanwhile.

File: docs/man-openmpi/man3/MPI_Request_get_status_all.3.rst (hunk 8 above)
Comment (edgargabriel) [reply docs/man-openmpi/man3/MPI_Request_get_status_all.3.rst]: My reading of the spec is that its not disallowed to call these functions with `MPI_STATUSES_IGNORE`. We are also allowing that already in `MPI_Request_get_status` today. There could be scenarios where it even makes sense, if the user is only interested in the `flag` without impacting the requests itself. The code will have to call `MPI_Testall` (or similar functions) anyway and could retrieve the actual status value at that time.

File: docs/man-openmpi/man3/MPI_Request_get_status_all.3.rst [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+
+.. include_body
+
+:ref:`MPI_Request_get_status_all` |mdash| Access information associated with a
+request without freeing the request.
+
+.. The following file was automatically generated
+.. include:: ./bindings/mpi_request_get_status_all.rst
+
+INPUT PARAMETER
```
Comment (jsquyres) [inline docs/man-openmpi/man3/MPI_Request_get_status_all.3.rst]: PARAMETERS (plural)

File: docs/man-openmpi/man3/MPI_Request_get_status_any.3.rst [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+
+.. include_body
+
+:ref:`MPI_Request_get_status_any` |mdash| Access information associated with a
+request without freeing the request.
+
+.. The following file was automatically generated
+.. include:: ./bindings/mpi_request_get_status_any.rst
+
+INPUT PARAMETER
```
Comment (jsquyres) [inline docs/man-openmpi/man3/MPI_Request_get_status_any.3.rst]: PARAMETERS  (plural)

File: docs/man-openmpi/man3/MPI_Request_get_status_some.3.rst [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+
+.. include_body
+
+:ref:`MPI_Request_get_status_some` |mdash| Access information associated with a
+request without freeing the request.
+
+.. The following file was automatically generated
+.. include:: ./bindings/mpi_request_get_status_some.rst
+
+INPUT PARAMETER
```
Comment (jsquyres) [inline docs/man-openmpi/man3/MPI_Request_get_status_some.3.rst]: PARAMETERS (plural)

File: docs/man-openmpi/man3/MPI_Request_get_status_all.3.rst (hunk 6 above)
Comment (jsquyres) [reply docs/man-openmpi/man3/MPI_Request_get_status_all.3.rst]: This still needs to be updated -- it's an array of requests, not a single request.  Perhaps:

> Non-destructively access status information associated with an array of requests

File: docs/man-openmpi/man3/MPI_Request_get_status_all.3.rst [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+any of the MPI test, wait, or free routines should be executed with each
+of those requests.
+
+Each status entry that corresponds to a null or inactive handle is set
+to empty.  Otherwise, ``flag = false`` is returned and the values of the
+status entries are undefined.
+
+If your application does not need to examine the *status* field, you can
+save resources by using the predefined constant ``MPI_STATUSES_IGNORE`` as a
+special value for the ``array_of_statuses`` argument.
```
Comment (jsquyres) [inline docs/man-openmpi/man3/MPI_Request_get_status_all.3.rst:43]: Although this may be allowed (per your prior response), I don't think we should say this.

If this text is straight out of the standard, we might want to quote it and say that it's allowed, but it's also a waste to of CPU cycles to call this function with `MPI_STATUSES_IGNORE`, because -- by definition -- it will effectively be a no-op.

File: ompi/mpi/c/request_get_status_all.c.in (hunk 10 above)
Comment (jsquyres) [reply ompi/mpi/c/request_get_status_all.c.in:61]: Ok.  Can you put a comment in about that?

File: ompi/mpi/fortran/mpif-h/request_get_status_some_f.c (hunk 13 above)
Comment (jsquyres) [reply ompi/mpi/fortran/mpif-h/request_get_status_some_f.c]: It looks like this was not fixed...?

File: ompi/mpi/fortran/use-mpi-f08/request_get_status_all.c.in (hunk 11 above)
Comment (jsquyres) [reply ompi/mpi/fortran/use-mpi-f08/request_get_status_all.c.in:57]: Might be worth a comment here explaining this -- future readers of the C code here might get confused on this point, too.

File: ompi/mpi/fortran/use-mpi-f08/request_get_status_any.c.in (hunk 12 above)
Comment (jsquyres) [reply ompi/mpi/fortran/use-mpi-f08/request_get_status_any.c.in:61]: Probably worth a comment -- this is a subtle point.

File: docs/man-openmpi/man3/MPI_Request_get_status_all.3.rst (hunk 17 above)
Comment (edgargabriel) [reply docs/man-openmpi/man3/MPI_Request_get_status_all.3.rst:43]: no, that is not correct. The flag indicating whether the operations have completed or not is being set, so its not a no-op

Comment (jsquyres) [reply docs/man-openmpi/man3/MPI_Request_get_status_all.3.rst:43]: Ooohhh... good point.  Ok, I retract my comments about this.

File: ompi/mpi/fortran/use-mpi-f08/request_get_status_all.c.in (hunk 11 above)
Comment (jsquyres) [reply ompi/mpi/fortran/use-mpi-f08/request_get_status_all.c.in:57]: The more I think about this, the more I worry about future reading / maintenance of the code where the Fortran binding is a LOGICAL value, but we're treating it as a C int.

I say this because anyone reading the Fortran bindings will be confused, like I was, and think, "oh no! that value is supposed to be a Fortran LOGICAL, not INTEGER!" and then go down a rabbit hole of investigation before learning (or remembering) that it's actually ok because our generated bindings are actually swizzling the LOGICAL to a C int appropriately.

Put simply: do we want to have to keep explaining this point every time the issue comes up?

I say this because it's a pretty subtle point, and not obvious at all when reading the C code of the Fortran bindings.  My $0.02 is that it will definitely come up in the future when new people read this code.  If that's a valid assumption, it might be better to improve this now and avoid developer confusion in the future.  I can think of 2 ways to improve this:

1. Put a comment everywhere that this happens.  It can be a short comment, just referring to a longer explanation elsewhere.
2. Have LOGICAL-named macros that really just do INTEGER things under the covers.  Also have appropriate comments in the `#define` implementations of those macros.  Hence, the C code here will read well/easily, and if someone goes to look into the `#define` implementations, they'll see the LOGICAL<-->C int swizzling and comments explaining why/what is happening.

But perhaps there are other ways, too -- I'm not tied to these 2 options.

Comment (hppritcha) [reply ompi/mpi/fortran/use-mpi-f08/request_get_status_all.c.in:57]: Okay probably some variant of 2 would make most sense.  I would rather do this kind of change in a separate PR as it will impact a bunch of files having nothing to do with this feature.

Comment (jsquyres) [reply ompi/mpi/fortran/use-mpi-f08/request_get_status_all.c.in:57]: That makes sense; we don't need to mix a global Fortran-generation-code stuff here in this PR.

**Comment by hppritcha** [comment]
did we add the tests for this to ompi-tests?  i'm triaging compile failures for the tests in my mtt runs:
```

request_get_status_all.c:32:9: error: call to undeclared function 'MPI_Request_get_status_all'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]    32 \|         MPI_Request_get_status_all(1, req, &flag, statuses);       \|         ^ request_get_status_all.c:32:9: note: did you mean 'MPI_Request_get_status'? /home/howardp/mtt/master_scratch/MiddlewareBuild_OMPIMaster/include/mpi.h:2234:20: note: 'MPI_Request_get_status' declared here  2234 \| OMPI_DECLSPEC  int MPI_Request_get_status(MPI_Request request, int *flag,       \|                    ^ 1 error generated. make[1]: Leaving directory '/home/howardp/mtt/master_scratch/TestGet_IBM/ompi-tests/ibm/pt2pt' make[1]: *** [Makefile:1582: request_get_status_all.o] Error 1 make: *** [Makefile:430: all-recursive] Error 1
--
Environment |  

```

**Comment by edgargabriel** [comment]
> did we add the tests for this to ompi-tests? i'm triaging compile failures for the tests in my mtt runs:
> 
> ```
> 
> request_get_status_all.c:32:9: error: call to undeclared function 'MPI_Request_get_status_all'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]    32 \|         MPI_Request_get_status_all(1, req, &flag, statuses);       \|         ^ request_get_status_all.c:32:9: note: did you mean 'MPI_Request_get_status'? /home/howardp/mtt/master_scratch/MiddlewareBuild_OMPIMaster/include/mpi.h:2234:20: note: 'MPI_Request_get_status' declared here  2234 \| OMPI_DECLSPEC  int MPI_Request_get_status(MPI_Request request, int *flag,       \|                    ^ 1 error generated. make[1]: Leaving directory '/home/howardp/mtt/master_scratch/TestGet_IBM/ompi-tests/ibm/pt2pt' make[1]: *** [Makefile:1582: request_get_status_all.o] Error 1 make: *** [Makefile:430: all-recursive] Error 1
> --
> Environment |  
> ```

@hppritcha I added these tests to the ibm testsuite https://github.com/open-mpi/ompi-tests/pull/189

**Comment by edgargabriel** [comment]
It was probably too early to merge it, should have waited until we merge the actual code. Sorry, I didn't think it through :-(

**Comment by hppritcha** [comment]
related to #13279

File: ompi/mpi/fortran/mpif-h/request_get_status_some_f.c (hunk 13 above)
Comment (hppritcha) [reply ompi/mpi/fortran/mpif-h/request_get_status_some_f.c]: I think it was fixed.  now use *outcount at line 119.

File: ompi/include/mpi.h.in [hunk 18]
```diff
@@ -2227,6 +2227,15 @@ OMPI_DECLSPEC  MPI_Request MPI_Request_f2c(MPI_Fint request);
 OMPI_DECLSPEC  int MPI_Request_free(MPI_Request *request);
 OMPI_DECLSPEC  int MPI_Request_get_status(MPI_Request request, int *flag,
                                           MPI_Status *status);
+/* should be 'const MPI_Request array_of_requests[]' */
```
Comment (bosilca) [inline ompi/include/mpi.h.in]: ok ? It actually is marked as `const` !

File: ompi/mpi/c/request_get_status_all.c.in [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+
+#include "ompi/mpi/c/bindings.h"
+#include "ompi/runtime/params.h"
+#include "ompi/communicator/communicator.h"
+#include "ompi/errhandler/errhandler.h"
+#include "ompi/request/request.h"
+#include "ompi/request/grequest.h"
+#include "ompi/memchecker.h"
+
+/* Non blocking test for the request status. Upon completion, the request will
```
Comment (bosilca) [inline ompi/mpi/c/request_get_status_all.c.in]: This is not technically a `test` but a non blocking status retrieval.

File: ompi/mpi/c/request_get_status_all.c.in [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+
+        OMPI_ERR_INIT_FINALIZE(FUNC_NAME);
+        if (NULL == requests) {
+            rc = MPI_ERR_REQUEST;
+        } else {
+            /* Need to cast the const away, since the request argument for most
+            ** MPI functions invoking this functionality does not have the const
+            ** argument
+            */
+            if(!ompi_request_check_same_instance((MPI_Request *)requests, count) ) {
```
Comment (bosilca) [inline ompi/mpi/c/request_get_status_all.c.in:61]: Changing `ompi_request_check_same_instance` to use const make sense as part of this PR because it is the first time it is useful to have const in the API instead of casting everywhere.

File: ompi/mpi/c/request_get_status_all.c.in [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+
+    opal_atomic_mb();
+    int i;
+    all_done = true;
+    for (i = 0; i < count; i++) {
+        if( (requests[i] == MPI_REQUEST_NULL) || (requests[i]->req_state == OMPI_REQUEST_INACTIVE) ||
+            (requests[i]->req_complete) ) {
+            continue;
+        }
+	all_done = false;
```
Comment (bosilca) [inline ompi/mpi/c/request_get_status_all.c.in]: there are plenty if incorrect indentations (mix between tab and spaces). Please fix before merge.

File: ompi/mpi/c/request_get_status_all.c.in [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+        return MPI_SUCCESS;
+    }
+
+    for (i = 0; i < count; i++) {
+        if( (requests[i] == MPI_REQUEST_NULL) || (requests[i]->req_state == OMPI_REQUEST_INACTIVE) ) {
+            if (MPI_STATUS_IGNORE != statuses) {
+                OMPI_COPY_STATUS(&statuses[i], ompi_status_empty, false);
+            }
+        }
+        if (requests[i]->req_complete ) {
```
Comment (bosilca) [inline ompi/mpi/c/request_get_status_all.c.in]: Why is this test necessary ? The requirement to get to this code is that all requests are completed, which means that if they are not NULL or inactive they must be completed.

File: ompi/mpi/c/request_get_status_any.c.in [hunk 23]
```diff
@@ -0,0 +1,136 @@
...
+            }
+            return MPI_SUCCESS;
+        } else {
+            /* regular request but not complete */
+            all_inactive = false;
+        }
+    }
+
+    if (all_inactive) {
+	*flag = true;
```
Comment (bosilca) [inline ompi/mpi/c/request_get_status_any.c.in]: indentation.

File: ompi/mpi/fortran/mpif-h/request_get_status_all_f.c [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+    c_status = (MPI_Status*) (c_req + OMPI_FINT_2_INT(*count));
+
+    for (i = 0; i < OMPI_FINT_2_INT(*count); ++i) {
+        c_req[i] = PMPI_Request_f2c(array_of_requests[i]);
+    }
+
+    c_ierr = PMPI_Request_get_status_all(OMPI_FINT_2_INT(*count), c_req, OMPI_LOGICAL_SINGLE_NAME_CONVERT(flag), c_status);
+    if (NULL != ierr) *ierr = OMPI_INT_2_FINT(c_ierr);
+
+    if (MPI_SUCCESS == c_ierr) {
```
Comment (bosilca) [inline ompi/mpi/fortran/mpif-h/request_get_status_all_f.c:111]: If I understand this function correctly conversion only makes sense if flag is true.

File: ompi/mpi/fortran/mpif-h/request_get_status_any_f.c [hunk 25]
```diff
@@ -0,0 +1,129 @@
...
+
+        /* Increment index by one for fortran conventions */
+
+        OMPI_SINGLE_INT_2_FINT(indx);
+        if (MPI_UNDEFINED != *(OMPI_SINGLE_NAME_CONVERT(indx))) {
+            array_of_requests[OMPI_INT_2_FINT(*indx)] =
+                c_req[OMPI_INT_2_FINT(*indx)]->req_f_to_c_index;
+            ++(*indx);
+        }
+        if (!OMPI_IS_FORTRAN_STATUS_IGNORE(status)) {
```
Comment (bosilca) [inline ompi/mpi/fortran/mpif-h/request_get_status_any_f.c]: this code must go in the `if` above.

File: ompi/mpi/fortran/mpif-h/request_get_status_some_f.c [hunk 26]
```diff
@@ -0,0 +1,140 @@
...
+                                          c_status);
+    if (NULL != ierr) *ierr = OMPI_INT_2_FINT(c_ierr);
+
+    if (MPI_SUCCESS == c_ierr) {
+        OMPI_SINGLE_INT_2_FINT(*outcount);
+        OMPI_ARRAY_INT_2_FINT(array_of_indices, *outcount);
+
+        /* Increment indexes by one for fortran conventions */
+
+        if (MPI_UNDEFINED != OMPI_FINT_2_INT(*outcount)) {
```
Comment (bosilca) [inline ompi/mpi/fortran/mpif-h/request_get_status_some_f.c:121]: Are you sure this is correct ? First you alter the `array_of_requests` which is technically an IN. Second, you populate the `array_of_statuses` according to `incount` and not to `outcount`. Here is what the MPI standard states about this:
> Returns in the first outcount locations of the array array_of_indices the indices of these operations within the array array_of_requests; the array is indexed from zero in C and from one in Fortran. Returns in the first outcount locations of the array array_of_statuses the status for these completed operations.

File: ompi/mpi/fortran/mpif-h/request_get_status_any_f.c (hunk 25 above)
Comment (hppritcha) [reply ompi/mpi/fortran/mpif-h/request_get_status_any_f.c]: yep done.

File: ompi/mpi/fortran/mpif-h/request_get_status_all_f.c (hunk 24 above)
Comment (hppritcha) [reply ompi/mpi/fortran/mpif-h/request_get_status_all_f.c:111]: you're right.  fixed with extra commit.

File: ompi/mpi/fortran/mpif-h/request_get_status_some_f.c (hunk 26 above)
Comment (hppritcha) [reply ompi/mpi/fortran/mpif-h/request_get_status_some_f.c:121]: you're right here too.  fixed with extra commit

File: ompi/mpi/c/request_get_status_all.c.in (hunk 20 above)
Comment (edgargabriel) [reply ompi/mpi/c/request_get_status_all.c.in:61]: @bosilca just to clarify: If I would change the `ompi_request_check_same_instance` function to take `const MPI_Request *` as an argument, I would have to add typecasts to `MPI_Waitall/Waitany/Waitsome/Testall/Testany/Testsome` since they are not `const MPI_Request*`. Is that what you are suggesting?

My line of thinking was that even if we decide to make that change (which I don't think I am technically in favor), I would prefer to do that in a separate PR, since I didn't want to touch the Waitall/Testall and friends functions in this one.

Comment (bosilca) [reply ompi/mpi/c/request_get_status_all.c.in:61]: I am not sure I understand why you would have to make changes anywhere else? Passing a `non-const` to a `const` function argument does not require casting (unlike passing a `const` to a `non-const`).

My concern is that delayed things have a tendency to remain delayed and then forgotten because there is always something more important. This is a general statement, I'm not pointing fingers.

File: ompi/include/mpi.h.in (hunk 18 above)
Comment (edgargabriel) [reply ompi/include/mpi.h.in]: This is a remainder from the first commit when the interface generation could not handle `const MPI_Request*` yet. Will remove them.

File: ompi/mpi/c/request_get_status_all.c.in (hunk 20 above)
Comment (edgargabriel) [reply ompi/mpi/c/request_get_status_all.c.in:61]: I don't think that is correct. I made change that you suggested, and now I get a lot of warnings from the files that I mentioned:

```
testall_generated.c: In function 'PMPI_Testall':
testall_generated.c:64:50: warning: passing argument 1 of 'ompi_request_check_same_instance' from incompatible pointer type [-Wincompatible-pointer-types]
   64 |             if(!ompi_request_check_same_instance(requests, count) ) {
      |                                                  ^~~~~~~~
      |                                                  |
      |                                                  struct ompi_request_t **
In file included from ../../../ompi/communicator/comm_request.h:18,
                 from ../../../ompi/communicator/communicator.h:45,
                 from testall_generated.c:33:
../../../ompi/request/request.h:575:62: note: expected 'const ompi_request_t **' but argument is of type 'struct ompi_request_t **'
  575 | bool ompi_request_check_same_instance(const ompi_request_t** requests,
      |                                       ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
```

File: ompi/mpi/c/request_get_status_all.c.in (hunk 19 above)
Comment (edgargabriel) [reply ompi/mpi/c/request_get_status_all.c.in]: fixed, thanks!

File: ompi/mpi/c/request_get_status_all.c.in (hunk 21 above)
Comment (edgargabriel) [reply ompi/mpi/c/request_get_status_all.c.in]: fixed, re-untabified all three files.

File: ompi/mpi/c/request_get_status_all.c.in (hunk 22 above)
Comment (edgargabriel) [reply ompi/mpi/c/request_get_status_all.c.in]: removed the check

File: ompi/mpi/c/request_get_status_any.c.in (hunk 23 above)
Comment (edgargabriel) [reply ompi/mpi/c/request_get_status_any.c.in]: fixed.

File: ompi/mpi/c/request_get_status_all.c.in (hunk 20 above)
Comment (edgargabriel) [reply ompi/mpi/c/request_get_status_all.c.in:61]: @bosilca, I incorporated all of your change requests except for this const thing, can you please re-check and let me know how you would like to proceed.

File: ompi/mpi/c/request_get_status_any.c.in [hunk 27]
```diff
@@ -0,0 +1,136 @@
...
+               the query function to get the status (MPI-2:8.2), even if
+               the user passed STATUS_IGNORE. */
+            if (OMPI_REQUEST_GEN == requests[i]->req_type) {
+                ompi_grequest_invoke_query(requests[i], &requests[i]->req_status);
+            }
+            if (MPI_STATUS_IGNORE != status) {
+                OMPI_COPY_STATUS(status, requests[i]->req_status, false);
+            }
+            return MPI_SUCCESS;
+        } else {
```
Comment (bosilca) [inline ompi/mpi/c/request_get_status_any.c.in:108]: The `else` itself is unnecessary, it can be assumed from the rest of the checks. That's how it is done in the `some` function.

File: ompi/mpi/c/request_get_status_all.c.in (hunk 20 above)
Comment (bosilca) [reply ompi/mpi/c/request_get_status_all.c.in:61]: I'm puzzled by the error message as I don't see why we cannot pass a non-const argument to a const function. Anyway, as is this PR is good to go.

**Review by bosilca — APPROVED** [review]
